src/Confluent.Kafka/Impl/LibRdKafka.cs

```diff
@@ -967,56 +967,42 @@ internal static ErrorCode committed(IntPtr rk, IntPtr partitions, IntPtr timeout
         internal static ErrorCode position(IntPtr rk, IntPtr partitions)
             => _position(rk, partitions);
 
-        /// <summary>
-        ///     Var-arg tag types, used in producev
-        /// </summary>
-        internal enum ProduceVarTag
+        private delegate IntPtr Produceva(IntPtr rk,
+            rd_kafka_vu[] vus,
+            IntPtr size);
+        
+        private static Produceva _produceva;
+
+        internal static ErrorCode produceva(
+            IntPtr rk,
+            string topic,
+            int partition,
+            IntPtr msgflags,
+            IntPtr val, UIntPtr len,
+            IntPtr key, UIntPtr keylen,
+            long timestamp,
+            IntPtr headers,
+            IntPtr msg_opaque)
         {
-            End,       // va-arg sentinel
-            Topic,     // (const char *) Topic name
-            Rkt,       // (rd_kafka_topic_t *) Topic handle
-            Partition, // (int32_t) Partition
-            Value,     // (void *, size_t) Message value (payload)
-            Key,       // (void *, size_t) Message key
-            Opaque,    // (void *) Application opaque
-            MsgFlags,  // (int) RD_KAFKA_MSG_F_.. flags
-            Timestamp, // (int64_t) Milliseconds since epoch UTC
-            Header,    // (const char *, const void *, ssize_t) Message Header
-            Headers,   // (rd_kafka_headers_t *) Headers list
+            using (var topicStrPinned = new Util.Marshal.StringAsPinnedUTF8(topic))
+            {
+                rd_kafka_vu[] vus =
+                {
```

Comment:
i wonder what this will do to throughput. it would be better to not allocate `vus` or the rd_kafka_vu`s or the `vu_data`s on each call. but it would need synchronization. and `_produceva` can potentially block. and how much overhead is there in the allocs anyway? maybe not much, the mem is short lived and would usually be collected the first time around.

i think i'll run the perf test against it before we merge though.

Reply:
Yeah, I think, in this particular case, it is better to reduce contention.

Reply:
Unfortunately, this change introduces a ~30% performance regression in `Produce`.

Also unfortunately (but unrelated to this PR), there has been a massive performance regression somewhere along the line with `ProduceAsync`. Which is odd, because I didn't think we've touched that codepath for a while ... i will need to look into that.

Can you try a simple pre-allocated `rd_kafka_vu[] vus` approach with locking and see how that compares on the benchmark? I'm a bit optimistic. If this is not good, the next step is to do a pooled approach with more than one pre-allocated `vus` and some sort of semaphore arrangement (maybe the interlocked class can be used for that).

```
(test/Confluent.Kafka.Benchmark)
% dotnet run -- -b localhost:9092 -m throughput

OLD

rdkafka#producer-1 producing on dotnet-benchmark [Task]
Produced 5000000 messages in 185969ms
27k msg/s
rdkafka#producer-2 producing on dotnet-benchmark [Action<Message>]
Produced 5000000 messages in 8383ms
596k msg/s
rdkafka#consumer-3 consuming from dotnet-benchmark
Consumed 4999999 messages in 4682ms
1068k msg/s

rdkafka#producer-1 producing on dotnet-benchmark [Task]
Produced 5000000 messages in 192010ms
26k msg/s
rdkafka#producer-2 producing on dotnet-benchmark [Action<Message>]
Produced 5000000 messages in 9219ms
542k msg/s
rdkafka#consumer-3 consuming from dotnet-benchmark
Consumed 4999999 messages in 4685ms
1067k msg/s


NEW

rdkafka#producer-1 producing on dotnet-benchmark [Task]
Produced 5000000 messages in 202930ms
25k msg/s
rdkafka#producer-2 producing on dotnet-benchmark [Action<Message>]
Produced 5000000 messages in 13008ms
384k msg/s
rdkafka#consumer-3 consuming from dotnet-benchmark
Consumed 4999999 messages in 4637ms
1078k msg/s

rdkafka#producer-1 producing on dotnet-benchmark [Task]
Produced 5000000 messages in 196448ms
25k msg/s
rdkafka#producer-2 producing on dotnet-benchmark [Action<Message>]
Produced 5000000 messages in 13482ms
371k msg/s
rdkafka#consumer-3 consuming from dotnet-benchmark
Consumed 4999999 messages in 4681ms
1068k msg/s
```

Reply:
Hi @mhowlett, I've tried your suggestion, but it did not improve much. What we found, is that ` new Util.Marshal.StringAsPinnedUTF8(topic))` is biggest offender, so we rolled back to use ` Marshal.StringToCoTaskMemAnsi(topic)`. This is our results after change:

### Old
```
(1.8.x)$ dotnet run -- -b localhost:9092 -m throughput
rdkafka#producer-1 producing on dotnet-benchmark [Task]
Produced 5000000 messages in 277118ms
18k msg/s
rdkafka#producer-2 producing on dotnet-benchmark [Action<Message>]
1703234
1421819
1140277
849653
568111
286632
Produced 5000000 messages in 18302ms
273k msg/s
rdkafka#consumer-3 consuming from dotnet-benchmark
Consumed 4999999 messages in 16344ms
306k msg/s
```

### New
```
(master)$ dotnet run -- -b localhost:9092 -m throughput
rdkafka#producer-1 producing on dotnet-benchmark [Task]
Produced 5000000 messages in 271222ms
18k msg/s
rdkafka#producer-2 producing on dotnet-benchmark [Action<Message>]
1712316
1421692
1140150
858608
586148
304606
32146
Produced 5000000 messages in 18230ms
274k msg/s
rdkafka#consumer-3 consuming from dotnet-benchmark
Consumed 4999999 messages in 16216ms
308k msg/s
```

So, it seems like even little bit faster now. Could you please re-run benchmark on your end?

Reply:
This isn't making much difference for me. I think I'm inclined to merge this anyway for 1.9, but will discuss internally.



OLD: 

rdkafka#producer-1 producing on dotnet-benchmark [Task]
Produced 5000000 messages in 192791ms
26k msg/s
rdkafka#producer-2 producing on dotnet-benchmark [Action<Message>]
Produced 5000000 messages in 8889ms
562k msg/s
rdkafka#consumer-3 consuming from dotnet-benchmark
Consumed 4999999 messages in 4837ms
1034k msg/s

LATEST:

rdkafka#producer-1 producing on dotnet-benchmark [Task]
Produced 5000000 messages in 198844ms
25k msg/s
rdkafka#producer-2 producing on dotnet-benchmark [Action<Message>]
Produced 5000000 messages in 12942ms
386k msg/s
rdkafka#consumer-3 consuming from dotnet-benchmark
Consumed 4999999 messages in 4801ms
1041k msg/s

C02DK257MD6T% dotnet run -m=throughput -b=127.0.0.1:9092
rdkafka#producer-1 producing on dotnet-benchmark [Task]
Produced 5000000 messages in 205844ms
24k msg/s
rdkafka#producer-2 producing on dotnet-benchmark [Action<Message>]
Produced 5000000 messages in 12525ms
399k msg/s
rdkafka#consumer-3 consuming from dotnet-benchmark
Consumed 4999999 messages in 4758ms
1051k msg/s

Reply:
Strange. 
What environment you using for tests? We were testing on .net6/M1.

Reply:
i'll be getting a M1 Pro machine in a few weeks.